src/shared_types.rs

```diff
@@ -0,0 +1,283 @@
+use serde::de;
+use serde::de::Deserializer;
+use serde::ser;
+use serde::ser::{SerializeStruct, Serializer};
+use serde::{Deserialize, Serialize};
+
+#[cfg(feature = "druid")]
+use druid::Data;
+
+/// Identifiers are optional attributes of several objects in the UFO.
+/// These identifiers are required to be unique within certain contexts
+/// as defined on a per object basis throughout this specification.
+/// Identifiers are specified as a string between one and 100 characters long.
+/// All characters must be in the printable ASCII range, 0x20 to 0x7E.
+#[derive(Debug, Clone, PartialEq)]
+pub struct Identifier(pub(crate) String);
+
+impl Identifier {
+    fn is_valid(&self) -> bool {
+        if self.0.len() > 100 {
+            return false;
+        }
+        for c in self.0.chars() {
+            if !(0x20..=0x7F).contains(&(c as u8)) {
+                return false;
+            }
+        }
+
+        true
+    }
+}
+
+impl Serialize for Identifier {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        if !self.is_valid() {
+            return Err(ser::Error::custom("Identifier must be at most 100 characters long and contain only ASCII characters in the range 0x20 to 0x7F."));
```

Comment:
Another way to express this is to skip the `return` and have an `else` block, so:

```rust
if self.is_valid() {
    serializer.serialize_str(&self.0)
} else {
    Err(MyErr)
}
```